docs/manpages/tracker-index.1
.TH tracker-index 1 "September 2014" GNU "User Commands"

.SH NAME
tracker-index \- List, pause, resume and command data miners indexing content

.SH SYNOPSIS
.nf
\fBtracker index\fR \-\-reindex\-mime\-type <\fImime1\fR> [[\-m [\fImime2\fR]] ...]
\fBtracker index\fR \-\-file <\fIfile1\fR> [[\fIfile2\fR] ...]
\fBtracker index\fR \-\-import <\fIfile1\fR> [[\fIfile2\fR] ...]
\fBtracker index\fR \-\-backup <\fIfile\fR> | \-\-restore <\fIfile\fR>
.fi

.SH DESCRIPTION
This command perform actions on the current index. The "index" holds a
snapshot of the working tree in a database.

The index command allows some level of control on existing data
indexed, such as re-indexing content from a specific demographic -
e.g. all JPEG images, or simply reindexing an existing or non-existent
file.

It may be a good idea to backup your index before an upgrade in case
there is data loss (which should never happen). In those cases, the
backup command is made available and of course the restore command
will import an older data set (or index) into an empty index.

Finally, there is an import feature which makes testing or applying a
"base" data set for use much easier.

.SH OPTIONS
.TP
.B \-m, \-\-reindex-mime\-type\fR=<\fImime1\fR> [[\-m [\fImime2\fR]] ...]
Re-index files which match the \fImime\fR type supplied. This is
usually used when installing new extractors which support \fImime\fR
types previously unsupported. This forces Tracker to re-index those
files. You can use \fB\-\-reindex\-mime\-type\fR more than once per
\fImime\fR type.
.TP
.B \-f, \-\-index\fR=<\fIfile1\fR> [[\fIfile2\fR] ...]
(Re)index a file matching the \fIfile\fR name(s) supplied.
.TP
.B \-b, \-\-backup\fR=<\fIfile\fR>
Begins backing up the Tracker databases and save it to the \fIfile\fR
given.
.TP
.B \-o, \-\-restore\fR=<\fIfile\fR>
Begins restoring a previous backup from the \fIfile\fR which points to
the location of the backup generated by \fB\-\-backup\fR.
.TP
.B \i, \-\-import\fR=<\fIfile1\fR> [[\fIfile2\fR] ...]
Allows data to be imported into the index / database by providing
files with Turtle content.

Multiple \fIfile\fR arguments can be provided to import data from
multiple files.

The \fIfile\fR argument can be either a local path or a URI. It also
does not have to be an absolute path.

.SH SEE ALSO
.BR tracker (1).
.BR Turtle.


NEWS
NEW in 1.2.6: 2015-04-09
========================

  * Fixes #746401, Fix possible maxCardinality constraint breaks in epub extractor
  * Fixes #746194, Ambiguous vala references
  * Fixes #735406, Restrict the amount of data that is logged for errors
  * Fixes #742035, Fix logic in guarantee title from file
  * Fixes #678986, Fix item_move() if source file does not exist

Translations: bs, cs, de, hu, id, lt, pl, pt_BR, sk, sl, sr, sr@latin, sv

NEW in 1.2.5: 2014-12-10
========================

  * Fixes: GB#741141, Tracker crashes every 5-10 minutes
  * Fixes: GB#733536, Remove --enable-guarantee metadata, always guarantee nie:title and nie:contentCreated
  * Fixes: GB#740268, System tab - very ambiguous about what is being removed
  * Fixes: GB#739149, Bump glib version (g_str_match_string is only in glib >= 2.40)
  * Fixes: RH#1133042, [abrt] tracker: persistence_store_file(): tracker-extract killed by SIGSEGV
  * libtracker-fts: Make property list per-thread data to fix uninitialisation issues
  * libtracker-fts: Fixed compilation errors when --disable-tracker-fts is used
  * tracker-extract: Fixed potential use of uninitialised hash table in persistence_invalidate()
  * tracker-extract: Fixed possible invalid memory read/write in tracker_keywords_parse()
  * tracker-extract: Fixed memory leak of MediaArt object for TrackerExtractInfo
  * clean-tracker-prefix: Should consider lib64 as well as lib directories

Translations:

  * Updated fi: Lasse Liehu, Jiri Grönroos
  * Updated pt: Duarte Loreto, Pedro Albuquerque


NEW in 1.2.4: 2014-11-06
========================

  * Fixes: GB#739237, Ownership reference not added in tracker-file-notifier.c

Translations:

  * Updated hu: Gábor Kelemen, Gabor Kelemen


NEW in 1.2.3: 2014-10-17
========================

  * Fixes: GB#737768, fs: TrackerCrawler can crash due to invalid container pointers
  * Fixes: GB#738522, libtracker-data: Guard against failure to create a FTS update statement
  * Fixes: GB#738450, libtracker-data: Guard against failure to open the database
  * libtracker-miner: _NO_STAT breaks in FileNotifier without a GFileInfo
  * libtracker-miner: Handle multiple nfo:belongsToContainer properties to 1 nie:url
  * tracker-writeback: Remove broken code to set file attributes
  * functional-tests: Clean up writebacktest base class
  * functional-tests: Clean up minertest base class
  * functional-tests: Remove IgnoreNextUpdate wrapper, nobody should be using this
  * functional-tests: Remove all remaining 'wait for miner to be idle' calls
  * functional-tests: Allow waiting on a specific property, and fix tests that need this
  * functional-tests: Await resource addition / removal properly in 300, 310, 600
  * functional-tests: Improve the wait-for-change code in helpers.StoreHelper
  * functional-tests: Partially fix writeback tests

Translations:

  * Updated it: Milo Casagrande
  * Updated sr: Мирослав Николић
  * Updated sr@latin: Мирослав Николић, Miroslav Nikolić


NEW in 1.2.2 - 2014-09-24
=========================

The main changes since 1.2.1 are:

  * Fixes: GB#737243, API break with tracker_indexing_tree_new()
  * libtracker-miner: Fixed missing Vala bindings for new DirectoryFlags.{PRIORITY|NO_STAT} enums


NEW in 1.2.1 - 2014-09-23
=========================

The main changes since 1.2.0 are:

  * ontology: Revert nfo:hasMediaStream max cardinality change until
    Philip merges his feature branch to allow database migration and
    coping for it. This won't be in the 1.2.x releases but likely in
    1.4. and master branch.

    For packagers, skipping 1.2.0 and using 1.2.1 should avoid data
    loss by way of a reindex that may be forced when users upgrade.
    This is highly important and something we want to avoid.


NEW in 1.2.0 - 2014-09-22
=========================

The main changes since 1.0.x are:

  * New Feature: Extractor now supports eBook and Comic book formats
  * New Feature: Extractor now supports BMP files
  * New Feature: Data Provider: Now there are TrackerDataProvider and TrackerEnumerator interfaces to allow 3rd parties to supply their own data to Tracker to be indexed, for example for an online or proprietary service
  * New Feature: Extractor is now 'crash-aware' that being, we have a separate data source for content that can't be extracted to avoid repeated failures from tracker-extract. This also survives through crashes that may occur due to attempted extractions
  * build: Updated doap file (added programming language and category is now "core")
  * build: Improve firefox and thunderbird detection to avoid warnings in console
  * build: Removed all GNU_SOURCE additions in sources, use AC_USE_SYSTEM_EXTENSIONS instead
  * build: Fixed many unit tests that were not properly isolated
  * build: configure now has an --enable-minimal option to ONLY build tracker-store
  * build: Reduce vala generated source warnings from GCC
  * build: Use major version only PkgConfig files. Previously, we had $library-$major-$minor (e.g. tracker-sparql-1.2, when it should be tracker-sparql-1.0) to avoid creating unnecessary work for developers using Tracker. This release corrects this and it was planned originally.
  * build: Include COPYING*GPL files in tarball
  * build: Generate Vala bindings from GIR for libtracker-miner, libtracker-control
  * build: Improved many GIR annotations for nullable cases in libtracker-miner, libtracker-control
  * build: Avoid using rpath where possible
  * build: Make check fails to generate test reports
  * build: Remove --enable-guarantee metadata, always guarantee nie:title and nie:contentCreated
  * build: Added CC_CHECK_FLAGS_APPEND m4 macros from xinelib/systemd projects to check GCC compiler flags we can use
  * ontology: nfo:hasMediaStream no longer has a max cardinality of 1, this may mean a reindex is needed
  * libtracker-bus: Fixed a crash in query_async() where splice async operation is cancelled

src/tracker/tracker-index.c
/*
 * Copyright (C) 2014, Lanedo <martyn@lanedo.com>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

#include "config.h"

#include <stdlib.h>
#include <errno.h>

#ifdef __sun
#include <procfs.h>
#endif

#include <glib.h>
#include <glib/gi18n.h>
#include <gio/gio.h>

#include <libtracker-control/tracker-control.h>
#include <libtracker-sparql/tracker-sparql.h>

#include "tracker-index.h"
#include "tracker-dbus.h"

static gchar **reindex_mime_types;
static gboolean index_file;
static gboolean backup;
static gboolean restore;
static gboolean import;
static gchar **filenames;

#define INDEX_OPTIONS_ENABLED()	  \
	((filenames && g_strv_length (filenames) > 0) || \
	 (index_file || \
	  backup || \
	  restore || \
	  import) || \
	 reindex_mime_types)

static GOptionEntry entries[] = {
	{ "reindex-mime-type", 'm', 0, G_OPTION_ARG_STRING_ARRAY, &reindex_mime_types,
	  N_("Tell miners to reindex files which match the mime type supplied (for new extractors), use -m MIME1 -m MIME2"),
	  N_("MIME") },
	{ "file", 'f', 0, G_OPTION_ARG_NONE, &index_file,
	  N_("Tell miners to (re)index a given file"),
	  N_("FILE") },
	{ "backup", 'b', 0, G_OPTION_ARG_NONE, &backup,
	  N_("Backup current index / database to the file provided"),
	  NULL },
	{ "restore", 'o', 0, G_OPTION_ARG_NONE, &restore,
	  N_("Restore a database from a previous backup (see --backup)"),
	  NULL },
	{ "import", 'i', 0, G_OPTION_ARG_NONE, &import,
	  N_("Import a dataset from the provided file (in Turtle format)"),
	  NULL },
	{ G_OPTION_REMAINING, 0, 0, G_OPTION_ARG_FILENAME_ARRAY, &filenames,
	  N_("FILE"),
	  N_("FILE") },
	{ NULL }
};

static gboolean
has_valid_uri_scheme (const gchar *uri)
{
	const gchar *s;

	s = uri;

	if (!g_ascii_isalpha (*s)) {
		return FALSE;
	}

	do {
		s++;
	} while (g_ascii_isalnum (*s) || *s == '+' || *s == '.' || *s == '-');

	return (*s == ':');
}

static gchar *
get_uri_from_arg (const gchar *arg)
{
	gchar *uri;

	/* support both, URIs and local file paths */
	if (has_valid_uri_scheme (arg)) {
		uri = g_strdup (arg);
	} else {
		GFile *file;

		file = g_file_new_for_commandline_arg (arg);
		uri = g_file_get_uri (file);
		g_object_unref (file);
	}

	return uri;
}

static int
reindex_mimes (void)
{
	GError *error = NULL;
	TrackerMinerManager *manager;

	/* Auto-start the miners here if we need to */
	manager = tracker_miner_manager_new_full (TRUE, &error);
	if (!manager) {
		g_printerr (_("Could not reindex mimetypes, manager could not be created, %s"),
		            error ? error->message : _("No error given"));
		g_printerr ("\n");
		g_clear_error (&error);
		return EXIT_FAILURE;
	}

	tracker_miner_manager_reindex_by_mimetype (manager, (GStrv) reindex_mime_types, &error);
	if (error) {
		g_printerr ("%s: %s\n",
		            _("Could not reindex mimetypes"),
		            error->message);
		g_error_free (error);
		return EXIT_FAILURE;
	}

	g_print ("%s\n", _("Reindexing mime types was successful"));
	g_object_unref (manager);

	return EXIT_SUCCESS;
}

static gint
index_or_reindex_file (void)
{
	TrackerMinerManager *manager;
	GError *error = NULL;
	gchar **p;

	/* Auto-start the miners here if we need to */
	manager = tracker_miner_manager_new_full (TRUE, &error);
	if (!manager) {
		g_printerr (_("Could not (re)index file, manager could not be created, %s"),
		            error ? error->message : _("No error given"));
		g_printerr ("\n");
		g_clear_error (&error);
		return EXIT_FAILURE;
	}

	for (p = filenames; *p; p++) {
		GFile *file;

		file = g_file_new_for_commandline_arg (*p);
		tracker_miner_manager_index_file (manager, file, &error);

		if (error) {
			g_printerr ("%s: %s\n",
			            _("Could not (re)index file"),
			            error->message);
			g_error_free (error);
			return EXIT_FAILURE;
		}

		g_print ("%s\n", _("(Re)indexing file was successful"));

		g_object_unref (manager);
		g_object_unref (file);
	}

	return EXIT_SUCCESS;
}

static int
import_turtle_files (void)
{
	TrackerSparqlConnection *connection;
	GError *error = NULL;
	gchar **p;

	connection = tracker_sparql_connection_get (NULL, &error);

	if (!connection) {
		g_printerr ("%s: %s\n",
		            _("Could not establish a connection to Tracker"),
		            error ? error->message : _("No error given"));
		g_clear_error (&error);
		return EXIT_FAILURE;
	}

	for (p = filenames; *p; p++) {
		GError *error = NULL;
		GFile *file;

		g_print ("%s:'%s'\n",
		         _("Importing Turtle file"),
		         *p);

		file = g_file_new_for_commandline_arg (*p);
		tracker_sparql_connection_load (connection, file, NULL, &error);
		g_object_unref (file);

		if (error) {
			g_printerr ("  %s, %s\n",
			            _("Unable to import Turtle file"),
			            error->message);

			g_error_free (error);
			continue;
		}

		g_print ("  %s\n", _("Done"));
		g_print ("\n");
	}

	g_object_unref (connection);

	return EXIT_SUCCESS;
}

static int
backup_index (void)
{
	GDBusConnection *connection;
	GDBusProxy *proxy;
	GError *error = NULL;
	GVariant *v;
	gchar *uri;

	if (!tracker_dbus_get_connection ("org.freedesktop.Tracker1",
	                                  "/org/freedesktop/Tracker1/Backup",
	                                  "org.freedesktop.Tracker1.Backup",
	                                  G_DBUS_PROXY_FLAGS_NONE,
	                                  &connection,
	                                  &proxy)) {
		return EXIT_FAILURE;
	}

	uri = get_uri_from_arg (filenames[0]);

	g_print ("%s\n", _("Backing up database"));
	g_print ("  %s\n", uri);

	/* Backup/Restore can take some time */
	g_dbus_proxy_set_default_timeout (proxy, G_MAXINT);

	v = g_dbus_proxy_call_sync (proxy,
	                            "Save",
	                            g_variant_new ("(s)", uri),
	                            G_DBUS_CALL_FLAGS_NONE,
	                            -1,
	                            NULL,
	                            &error);

	if (proxy) {
		g_object_unref (proxy);
	}

	if (error) {
		g_critical ("%s, %s",
		            _("Could not backup database"),
		            error ? error->message : _("No error given"));
		g_clear_error (&error);
		g_free (uri);

		return EXIT_FAILURE;
	}

	if (v) {
		g_variant_unref (v);
	}

	g_free (uri);

	return EXIT_SUCCESS;
}

static int
restore_index (void)
{
	GDBusConnection *connection;
	GDBusProxy *proxy;
	GError *error = NULL;
	GVariant *v;
	gchar *uri;

	if (!tracker_dbus_get_connection ("org.freedesktop.Tracker1",
	                                  "/org/freedesktop/Tracker1/Backup",
	                                  "org.freedesktop.Tracker1.Backup",
	                                  G_DBUS_PROXY_FLAGS_NONE,
	                                  &connection,
	                                  &proxy)) {
		return EXIT_FAILURE;
	}

	uri = get_uri_from_arg (filenames[0]);

	g_print ("%s\n", _("Restoring database from backup"));
	g_print ("  %s\n", uri);

	/* Backup/Restore can take some time */
	g_dbus_proxy_set_default_timeout (proxy, G_MAXINT);

	v = g_dbus_proxy_call_sync (proxy,
	                            "Restore",
	                            g_variant_new ("(s)", uri),
	                            G_DBUS_CALL_FLAGS_NONE,
	                            -1,
	                            NULL,
	                            &error);

	if (proxy) {
		g_object_unref (proxy);
	}

	if (error) {
		g_critical ("%s, %s",
		            _("Could not backup database"),
		            error ? error->message : _("No error given"));
		g_clear_error (&error);
		g_free (uri);

		return EXIT_FAILURE;
	}

	if (v) {
		g_variant_unref (v);
	}

	g_free (uri);

	return EXIT_SUCCESS;
}

static int
index_run (void)
{
	if (reindex_mime_types) {
		return reindex_mimes ();
	}

	if (index_file) {
		return index_or_reindex_file ();
	}

	if (import) {
		return import_turtle_files ();
	}

	if (backup) {
		return backup_index ();
	}

	if (restore) {
		return restore_index ();
	}

	/* All known options have their own exit points */
	g_warn_if_reached ();

	return EXIT_FAILURE;
}

static int
index_run_default (void)
{
	GOptionContext *context;
	gchar *help;

	context = g_option_context_new (NULL);
	g_option_context_add_main_entries (context, entries, NULL);
	help = g_option_context_get_help (context, TRUE, NULL);
	g_option_context_free (context);
	g_printerr ("%s\n", help);
	g_free (help);

	return EXIT_FAILURE;
}

static gboolean
index_options_enabled (void)
{
	return INDEX_OPTIONS_ENABLED ();
}

int
tracker_index (int argc, const char **argv)
{
	GOptionContext *context;
	GError *error = NULL;
	const gchar *failed;
	gint actions = 0;

	context = g_option_context_new (NULL);
	g_option_context_add_main_entries (context, entries, NULL);

	argv[0] = "tracker index";

	if (!g_option_context_parse (context, &argc, (char***) &argv, &error)) {
		g_printerr ("%s, %s\n", _("Unrecognized options"), error->message);
		g_error_free (error);
		g_option_context_free (context);
		return EXIT_FAILURE;
	}

	g_option_context_free (context);

	if (backup) {
		actions++;
	}

	if (restore) {
		actions++;
	}

	if (index_file) {
		actions++;
	}

	if (import) {
		actions++;
	}

	if (actions > 1) {
		failed = _("Only one action (--backup, --restore, --index-file or --import) can be used at a time");
	} else if (actions > 0 && (!filenames || g_strv_length (filenames) < 1)) {
		failed = _("Missing one or more files which are required");
	} else if ((backup || restore) && (filenames && g_strv_length (filenames) > 1)) {
		failed = _("Only one file can be used with --backup and --restore");
	} else if (actions > 0 && (reindex_mime_types && g_strv_length (reindex_mime_types) > 0)) {
		failed = _("Actions (--backup, --restore, --index-file and --import) can not be used with --reindex-mime-type");
	} else {
		failed = NULL;
	}

	if (failed) {
		g_printerr ("%s\n\n", failed);
		return EXIT_FAILURE;
	}

	if (index_options_enabled ()) {
		return index_run ();
	}

	return index_run_default ();
}
